### PR TITLE
[WIP] Modularity Change Review

### PR DESCRIPTION
**Pull request by aubreymasten**
_Thank you for submitting a pull request. This code is leveraged to monitor critical services. Before contributing, please read our [contributing guidelines](https://github.com/newrelic/newrelic-browser-agent/blob/main/CONTRIBUTING.md) and [code of conduct](https://github.com/newrelic/.github/blob/main/CODE_OF_CONDUCT.md)._

---

### Overview
Please describe the changes present in the pull request and if applicable, describe why the changes are needed.

### Related Github Issue
Please include a link to the related GitHub issue, if applicable

### Testing
The browser agent includes a suite of tests which should be used to
verify that your changes don't break existing functionality. These tests will run through Github Actions when:

 1) A pull request is made _and_ 
 2) The pull request is marked as ready for testing by the internal team. 

If applicable, please include details on how to test your changes.

---   
_More details on running your tests locally can be found in the
[CONTRIBUTING](https://github.com/newrelic/newrelic-browser-agent/blob/main/CONTRIBUTING.md) and [README](https://github.com/newrelic/newrelic-browser-agent/blob/main/README.md) docs._

### CODE REVIEW

File: cdn/.env [hunk 1]
```diff
@@ -0,0 +1,2 @@
+PATH=
```
Comment (aubreymasten) [inline cdn/.env:1]: What does this file do?

File: packages/browser-agent/.env [hunk 2]
```diff
@@ -0,0 +1 @@
+SUBPATH=
```
Comment (aubreymasten) [inline packages/browser-agent/.env:1]: Does this serve a purpose? Should we gitignore this?

File: cdn/.env (hunk 1 above)
Comment (metal-messiah) [reply cdn/.env:1]: I think nothing, remnant of early webpack inquiry, should be removed

File: tools/jil/server/asset-server.js [hunk 3]
```diff
@@ -531,12 +590,12 @@ class AssetServer extends BaseServer {
       })
     } else {
       rsp.writeHead(404)
-      rsp.write(`Failed to locate asset at '${assetPath}'`)
+      // rsp.write(`Failed to locate asset at '${assetPath}'`)
```
Comment (aubreymasten) [inline tools/jil/server/asset-server.js:593]: Should we uncomment this?

File: packages/browser-agent/.env (hunk 2 above)
Comment (metal-messiah) [reply packages/browser-agent/.env:1]: Removing this may have an impact but only on the NPM package build process

File: tools/jil/server/asset-server.js (hunk 3 above)
Comment (metal-messiah) [reply tools/jil/server/asset-server.js:593]: I'll leave that up to you

File: packages/browser-agent-core/features/ajax/instrument/index.js [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+    this.txSize = dataSize(opts.body) || 0
+  }
+
+  // we capture failed call as status 0, the actual error is ignored
+  // eslint-disable-next-line handle-callback-err
+  function onFetchDone(err, res) {
+    this.endTime = now()
+    if (!this.params) {
+      this.params = {}
+    }
```
Comment (aubreymasten) [inline packages/browser-agent-core/features/ajax/instrument/index.js:319]: A check was removed for Data URLs that I think should still exist: https://github.com/newrelic/newrelic-browser-agent/blob/e0addb7c19e52429cc17b50a4e6f01892b3c1fc1/feature/xhr/instrument/index.js#L268-L269

File: packages/browser-agent-core/features/ajax/instrument/index.js [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+    this.dt = new DT(this.agentIdentifier)
+
+    this.handler = (type, args, ctx, group) => handle(type, args, ctx, group, this.ee)
+    this.wrappedFetch = getWrappedFetch(this.ee)
+    wrapXhr(this.ee)
+    subscribeToEvents(this.agentIdentifier, this.ee, this.handler, this.dt)
+  }
+}
+
+// TODO update all of this to go into class and use this.ee for ee, handle, and register
```
Comment (aubreymasten) [inline packages/browser-agent-core/features/ajax/instrument/index.js:44]: Do we want to take action on this before the release?

File: packages/browser-agent-core/features/ajax/instrument/index.js [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+
+    // Don't instrument Chrome for iOS, it is buggy and acts like there are URL verification issues
+    if (!agentRuntime.xhrWrappable || agentRuntime.disabled) return
+
+    agentRuntime.features.xhr = true;   // declare that we are using xhr instrumentation
+
+    this.dt = new DT(this.agentIdentifier)
+
+    this.handler = (type, args, ctx, group) => handle(type, args, ctx, group, this.ee)
+    this.wrappedFetch = getWrappedFetch(this.ee)
```
Comment (aubreymasten) [inline packages/browser-agent-core/features/ajax/instrument/index.js:38]: I don't see this used anywhere in the agent. Is this from the original modularity prototype? Can we delete it?

File: packages/browser-agent-core/features/ajax/instrument/index.js (hunk 5 above)
Comment (aubreymasten) [inline packages/browser-agent-core/features/ajax/instrument/index.js:44]: Should we create a ticket for this or update before release?

File: packages/browser-agent-core/features/ajax/instrument/index.js (hunk 4 above)
Comment (aubreymasten) [inline packages/browser-agent-core/features/ajax/instrument/index.js:319]: This is missing a supportability metric from the original code:
https://github.com/newrelic/newrelic-browser-agent/blob/e0addb7c19e52429cc17b50a4e6f01892b3c1fc1/feature/xhr/instrument/index.js#L287-L290

File: packages/browser-agent-core/features/ajax/instrument/index.js [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+    var url
+    if (typeof target === 'string') {
+      url = target
+    } else if (typeof target === 'object' && target instanceof origRequest) {
+      url = target.url
+    } else if (window.URL && typeof target === 'object' && target instanceof URL) {
+      url = target.href
+    }
+    addUrl(this, url)
+
```
Comment (aubreymasten) [inline packages/browser-agent-core/features/ajax/instrument/index.js:305]: Missing a check for Data URL:
https://github.com/newrelic/newrelic-browser-agent/blob/e0addb7c19e52429cc17b50a4e6f01892b3c1fc1/feature/xhr/instrument/index.js#L268-L269

File: packages/browser-agent-core/features/ajax/aggregate/index.js [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+import { FeatureBase } from '../../../common/util/feature-base'
+
+export class Aggregate extends FeatureBase {
+  constructor(agentIdentifier, aggregator, externalFeatures = []) {
+    super(agentIdentifier, aggregator, externalFeatures)
+    let ajaxEvents = []
+    let spaAjaxEvents = {}
+    let sentAjaxEvents = []
+    let scheduler
+
```
Comment (aubreymasten) [inline packages/browser-agent-core/features/ajax/aggregate/index.js:22]: Missing runtime check for whether instrumentation was loaded:

https://github.com/newrelic/newrelic-browser-agent/blob/e0addb7c19e52429cc17b50a4e6f01892b3c1fc1/feature/xhr/aggregate/index.js#L26-L31

File: packages/browser-agent-core/features/ajax/aggregate/index.js [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+    register('xhr', storeXhr, undefined, this.ee)
+
+    if (allAjaxIsEnabled()) {
+      scheduler = new HarvestScheduler('events', {
+        onFinished: onEventsHarvestFinished,
+        getPayload: prepareHarvest
+      }, this)
+
+      externalFeatures.forEach(feat => {
```
Comment (aubreymasten) [inline packages/browser-agent-core/features/ajax/aggregate/index.js:60]: This is the metrics harvest, should live outside of allAjaxIsEnabled check

File: packages/browser-agent-core/features/jserrors/instrument/index.js [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    })
+
+    state.ee.on('internal-error', (e) => {
+      handle('ierr', [e, now(), true], undefined, undefined, state.ee)
+    })
+
+    // Declare that we are using err instrumentation
+    // require('./debug')
+
+    const prevOnError = window.onerror
```
Comment (aubreymasten) [inline packages/browser-agent-core/features/jserrors/instrument/index.js:60]: This seems to have changed the behavior compared to the original code. Potentially causing the original onErrorHandler to be called twice.

File: packages/browser-agent-core/features/jserrors/instrument/index.js [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+    window.onerror = (...args) => {
+      if (prevOnError) prevOnError(...args)
+      this.onerrorHandler(...args)
+      return false
+    }
+
+    try {
+      window.addEventListener('unhandledrejection', (e) => {
```
Comment (aubreymasten) [inline packages/browser-agent-core/features/jserrors/instrument/index.js:68]: What is this try/catch supposed to do and where did it come from? It looks like it was added in the Dev commit but I don't see it in the original code.

File: packages/browser-agent-core/features/jserrors/aggregate/index.js [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+    super(agentIdentifier, aggregator)
+    this.stackReported = {}
+    this.pageviewReported = {}
+    this.errorCache = {}
+    this.currentBody
+
+    this.errorOnPage = false
+
+    // this will need to change to match whatever ee we use in the instrument
+    this.ee.on('interactionSaved',  (interaction) => {
```
Comment (aubreymasten) [inline packages/browser-agent-core/features/jserrors/aggregate/index.js:32]: Could we make this a named class function and move it out of the constructor?

File: packages/browser-agent-core/features/page-view-timing/instrument/index.js [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+    this.pageHiddenTime
+    this.performanceObserver
+    this.lcpPerformanceObserver
+    this.clsPerformanceObserver
+    this.fiRecorded = false
+
+
+    // // NREUM.debug("initialize page-view-timing instrument!", agentIdentifier)
+
+    if (this.isEnabled()) {
```
Comment (aubreymasten) [inline packages/browser-agent-core/features/page-view-timing/instrument/index.js:25]: This should be handled by getEnabledFeatures in the loader files:
https://github.com/newrelic/newrelic-browser-agent/blob/7922cefe29918f7a608e6f7c613b8d1c28de897f/cdn/agent-loader/lite.js#L23-L26

File: packages/browser-agent-core/features/ajax/instrument/index.js (hunk 5 above)
Comment (metal-messiah) [reply packages/browser-agent-core/features/ajax/instrument/index.js:44]: I think this is comment can be deleted, as `this.ee` is effectively passed into the dependent methods.

Comment (metal-messiah) [reply packages/browser-agent-core/features/ajax/instrument/index.js:44]: https://github.com/newrelic/newrelic-browser-agent/pull/255#discussion_r931703594

File: packages/browser-agent-core/features/ajax/instrument/index.js (hunk 4 above)
Comment (metal-messiah) [reply packages/browser-agent-core/features/ajax/instrument/index.js:319]: Good catch. Needs to be added back.

Comment (metal-messiah) [reply packages/browser-agent-core/features/ajax/instrument/index.js:319]: https://github.com/newrelic/newrelic-browser-agent/projects/7#card-84572817

File: packages/browser-agent-core/features/ajax/instrument/index.js (hunk 7 above)
Comment (metal-messiah) [reply packages/browser-agent-core/features/ajax/instrument/index.js:305]: https://github.com/newrelic/newrelic-browser-agent/projects/7#card-84572817

File: packages/browser-agent-core/features/ajax/aggregate/index.js (hunk 8 above)
Comment (metal-messiah) [reply packages/browser-agent-core/features/ajax/aggregate/index.js:22]: The flag was needed before the modularity work, because features initialized as a side-effect of import.  Now we can be more direct when initializing the class instances.

File: packages/browser-agent-core/features/ajax/aggregate/index.js (hunk 9 above)
Comment (metal-messiah) [reply packages/browser-agent-core/features/ajax/aggregate/index.js:60]: This is a nice catch.  This could be a new ticket for after the release, to extend the metrics feature to do more than just supportability and remove the dependency from AJAX feature.

File: packages/browser-agent-core/features/jserrors/instrument/index.js (hunk 10 above)
Comment (metal-messiah) [reply packages/browser-agent-core/features/jserrors/instrument/index.js:60]: This is necessary for multiple instances of the agent on one page to have a page-level error handler. An HTML page starts with `window.onerror` as null by default, meaning it will only fire previous error handlers if they exist at the time of instrumenting that instance of the jserrors class (presumably by being set by one or more previous instances of the agent). 

This can be discussed further if necessary

File: packages/browser-agent-core/features/jserrors/instrument/index.js (hunk 11 above)
Comment (metal-messiah) [reply packages/browser-agent-core/features/jserrors/instrument/index.js:68]: I **think** this was added in the original NPM work, to catch unhandled promise rejections.  This is likely unnecessary now that all the `wrap` files have been implemented, but it is worth checking.

File: packages/browser-agent-core/features/jserrors/aggregate/index.js (hunk 12 above)
Comment (metal-messiah) [reply packages/browser-agent-core/features/jserrors/aggregate/index.js:32]: agreed

File: packages/browser-agent-core/features/page-view-timing/instrument/index.js (hunk 13 above)
Comment (metal-messiah) [reply packages/browser-agent-core/features/page-view-timing/instrument/index.js:25]: agreed, can be removed

File: packages/browser-agent-core/features/ajax/instrument/index.js (hunk 5 above)
Comment (metal-messiah) [reply packages/browser-agent-core/features/ajax/instrument/index.js:44]: up for discussion though

Comment (cwli24) [reply packages/browser-agent-core/features/ajax/instrument/index.js:44]: I added this; you can delete it if you want; at this point, all of the instrument & aggregate could be re-evaluated for rework or styling changes.

File: packages/browser-agent-core/common/aggregate/aggregator.js [hunk 14]
```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2020 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { SharedContext } from '../context/shared-context'
+import { mapOwn } from '../util/map-own'
+
+export class Aggregator extends SharedContext {
```
Comment (aubreymasten) [inline packages/browser-agent-core/common/aggregate/aggregator.js:9]: Some clarity in the names of these functions has been lost because we've named the methods what they were named in the original export. Maybe we can review this file to rename and re-add some old comments:

https://github.com/newrelic/newrelic-browser-agent/blob/487a282f92dd02a32b2b94d41bd7b718f09ff30c/agent/aggregator.js#L11-L15

Comment (cwli24) [reply packages/browser-agent-core/common/aggregate/aggregator.js:9]: Btw my 2 cents from working on tests: it did not make sense as reader for the code to rename a specialized function with a generalized version of its name. For example, don't expose "storeMetricUnderAgentEE" as "storeMetric". It's non-descript, vague, and potentially misleading. Expose methods as they are, or at least abbreviate them without losing context, e.g. "storeMetricAgtEE".

File: packages/browser-agent-core/common/window/load.js [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+const doc = win.document
+const ADD_EVENT_LISTENER = 'addEventListener'
+const ATTACH_EVENT = 'attachEvent'
+
+function checkState (cb) {
+  if (doc.readyState === 'complete') cb()
+}
+
+export function onWindowLoad(cb) {
+  checkState(cb)
```
Comment (aubreymasten) [inline packages/browser-agent-core/common/window/load.js:13]: This checkState will call the onWindowLoaded callback if the agent loads after the window.load and set the onload and load timings incorrectly.
https://github.com/newrelic/newrelic-browser-agent/blob/7e13aea02cc7ac928592bfaf71e5a6f0d18bac23/packages/browser-agent-core/features/page-view-event/instrument/index.js#L18-L28

Comment (aubreymasten) [reply packages/browser-agent-core/common/window/load.js:13]: The old agent is clearer that `measureWindowLoaded()` (`windowLoaded()` in the old agent) should only ever be called more than once:

https://github.com/newrelic/newrelic-browser-agent/blob/7e13aea02cc7ac928592bfaf71e5a6f0d18bac23/loader/index.js#L77-L79

File: packages/browser-agent-core/common/window/load.js [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+  if (doc[ADD_EVENT_LISTENER]) win[ADD_EVENT_LISTENER]('load', cb, eventListenerOpts(false))
+  else win[ATTACH_EVENT]('onload', cb)
+}
+
+export function onDOMContentLoaded(cb) {
+  checkState(cb)
```
Comment (aubreymasten) [inline packages/browser-agent-core/common/window/load.js:19]: I think the comment above applies here too, `doc.readyState === 'complete'` could be any point after DOMContentLoaded. If we get to create a register before that event, we should call it, but if the event has already occurred we shouldn't call the callback immediately because it relies on `now()` and assumes we are at the moment of DOMContentLoaded.

File: packages/browser-agent-core/common/wrap/index.js [hunk 17]
```diff
@@ -0,0 +1,46 @@
+
```
Comment (aubreymasten) [inline packages/browser-agent-core/common/wrap/index.js:1]: What does this file do?

File: packages/browser-agent-core/common/wrap/wrap-events.js [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+
+  function uniqueListener(args, obj) {
+    // Context for the listener is stored on itself.
+    return args[1]
+  }
+
+  return ee
+}
+
+export function scopedEE(sharedEE){
```
Comment (aubreymasten) [inline packages/browser-agent-core/common/wrap/wrap-events.js:81]: I only see this `export function scopedEE` pattern used as an import for one test, and I'm not sure what it's doing.

I see scopedEE imported in this test, but I think the test needs to be changed to import and call wrapEvents

https://github.com/newrelic/newrelic-browser-agent/blob/7e13aea02cc7ac928592bfaf71e5a6f0d18bac23/tests/browser/add-event-listener-options.browser.js#L8-L10

Is there any other use for this extra export on all the wrap-* files?

File: tests/browser/add-event-listener-options.browser.js [hunk 19]
```diff
@@ -4,11 +4,10 @@
...
 
 const jil = require('jil')
-let matcher = require('../../tools/jil/util/browser-matcher')
-let supported = matcher.withFeature('addEventListenerOptions')
 
-jil.browserTest('addEventListener options work when wrapped', supported, function (t) {
-  require('../../feature/wrap-events.js')
+jil.browserTest('addEventListener options work when wrapped', async function (t) {
+  const { scopedEE } = await import('../../packages/browser-agent-core/common/wrap/wrap-events');
+  scopedEE();
```
Comment (aubreymasten) [inline tests/browser/add-event-listener-options.browser.js:10]: I think this may want to be replaced with wrapEvents()

File: packages/browser-agent-core/common/wrap/wrap-history.js [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+  var object = window.history
+  if (prototype && prototype.pushState && prototype.replaceState) {
+    object = prototype
+  }
+  // log('wrap history')
+  wrapFn.inPlace(object, [ 'pushState', 'replaceState' ], '-')
+
+  return ee
+}
+3
```
Comment (aubreymasten) [inline packages/browser-agent-core/common/wrap/wrap-history.js:28]: Remove

File: packages/browser-agent-core/common/wrap/wrap-promise.js [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+  
+        return returnVal
+      }
+    })
+  
+    OriginalPromise.prototype['catch'] = function wrappedCatch (fn) {
+      return this.then(null, fn)
+    }
+  
+    Object.assign(OriginalPromise.prototype, {constructor: {value: WrappedPromise}})
```
Comment (aubreymasten) [inline packages/browser-agent-core/common/wrap/wrap-promise.js:66]: Why was this changed?

File: packages/browser-agent-core/common/wrap/wrap-xhr.js [hunk 22]
```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright 2020 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+// wrap-events patches XMLHttpRequest.prototype.addEventListener for us.
+import './wrap-events'
```
Comment (aubreymasten) [inline packages/browser-agent-core/common/wrap/wrap-xhr.js:7]: I don't think this will guarantee that events are wrapped. Other features call wrap-events, so for Lite, Pro, SPA this should work. But the Ajax feature on its own assumes that `import './wrap-events'` does the same thing as the previous `require('./wrap-events')` and auto-wraps.

To be safe we may want to call wrapEvents() around line 22.

This is a dependency that maybe we want to represent somewhere more unified as well.

File: packages/browser-agent-core/common/url/protocol.js [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+export const protocol = {
+  isFileProtocol: isFileProtocol,
+  supportabilityMetricSent: false
+}
+
+function isFileProtocol () {
+  let win = getWindow();
+  let isFile = !!(win.location && win.location.protocol && win.location.protocol === 'file:');
+  if (isFile) {
+    //metrics.recordSupportability('Generic/FileProtocol/Detected') -- may be implemented later? Probably make sure it's once per window
```
Comment (aubreymasten) [inline packages/browser-agent-core/common/url/protocol.js:16]: This should be re-implemented before release. We may also want to note this as a risk for multiple agent instances, like the comment says

File: packages/browser-agent-core/common/util/obfuscate.js [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+let recordedSupportability = false
+
+export class Obfuscator extends SharedContext {
+  constructor(parent) {
+    super(parent) // gets any allowed properties from the parent and stores them in `sharedContext`
+
+    /* Note: we may need these lines from legacy code -- (when passing in an optional metrics Instrument instance as argument)
+    if (shouldObfuscate()) metrics.recordSupportability('Generic/Obfuscate/Detected')
```
Comment (aubreymasten) [inline packages/browser-agent-core/common/util/obfuscate.js:17]: These should be re-implemented before release

File: packages/browser-agent-core/common/util/obfuscate.js [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+    for (var i = 0; i < rules.length; i++) {
+      var regex = rules[i].regex
+      var replacement = rules[i].replacement || '*'
+      obfuscated = obfuscated.replace(regex, replacement)
+    }
+    return obfuscated
+  }
+}
+
+// TO DO: this function should be inside the Obfuscator class since its context relates to agentID
```
Comment (aubreymasten) [inline packages/browser-agent-core/common/util/obfuscate.js:44]: Just for consistency we should finish this task

File: packages/browser-agent-core/features/spa/instrument/index.js [hunk 26]
```diff
@@ -0,0 +1,110 @@
...
+        const tracerEE = this.ee.get('tracer')
+        const jsonpEE = wrapJson(this.ee)
+        const promiseEE = wrapPromise(this.ee)
+        const eventsEE = wrapEvents(this.ee)
+        const timerEE = wrapTimer(this.ee)
+        const xhrEE = wrapXhr(this.ee)
+        const fetchEE = wrapFetch(this.ee)
+        const historyEE = wrapHistory(this.ee)
+        const mutationEE = wrapMutation(this.ee)
+
```
Comment (aubreymasten) [inline packages/browser-agent-core/features/spa/instrument/index.js:48]: Importing and running `xhr/instrument` was removed, which will cause problems if the Ajax feature is not required to be imported or named as an explicit dependency of SPA:

https://github.com/newrelic/newrelic-browser-agent/blob/e0addb7c19e52429cc17b50a4e6f01892b3c1fc1/feature/spa/instrument/index.js#L38

We should re-add this or create a dependency so we're at least warned if Ajax is ever disabled

File: packages/browser-agent-core/features/spa/aggregate/index.js [hunk 27]
```diff
@@ -0,0 +1,708 @@
...
+        interaction.root.attrs.firstPaint = paintMetrics['first-paint']
+        interaction.root.attrs.firstContentfulPaint = paintMetrics['first-contentful-paint']
+      }
+      baseEE.emit('interactionSaved', [interaction])
+      state.interactionsToHarvest.push(interaction)
+      scheduler.scheduleHarvest(0)
+
+    }
+
+    function isEnabled() {
```
Comment (aubreymasten) [inline packages/browser-agent-core/features/spa/aggregate/index.js:693]: Is this now handled by the CDN setup code?

File: packages/browser-agent-core/features/spa/aggregate/index.js [hunk 28]
```diff
@@ -0,0 +1,708 @@
...
+        }, 0)
+      }, undefined, baseEE)
+
+      register(INTERACTION_API + 'onEnd', function (t, cb) {
+        this.ixn.handlers.push(cb)
+      }, undefined, baseEE)
+
+      register('api-routeName', function (t, currentRouteName) {
+        state.lastSeenRouteName = currentRouteName
+        if (state.currentNode) state.currentNode[INTERACTION].setNewRoute(currentRouteName)
```
Comment (aubreymasten) [inline packages/browser-agent-core/features/spa/aggregate/index.js:584]: This looks like it was added post 1216, was this to address a bug found in testing?

File: cdn/agent-aggregator/util/featureDependencies.js [hunk 29]
```diff
@@ -0,0 +1,10 @@
+export function getFeatureDependencyNames(feature) {
+    switch(feature){
+        case 'ajax':
+            return ['jserrors']
+        case 'session-trace':
+            return ['ajax']
```
Comment (aubreymasten) [inline cdn/agent-aggregator/util/featureDependencies.js:6]: We should probably add a SPA depends on Ajax rule here

**Comment by metal-messiah** [comment]
> Any plans drawn to optimize using `splitChunks` ? I imagine the chunks be: a. base loader + {metrics, page-view-event, page-view-timings} instrument b. **each** feature instrument individually c. base aggregator d. {m, pve, pvt} together + **each** feature aggregate individually

Plan would probably be an outcome of spiking on async loader impacts

File: packages/browser-agent-core/common/wrap/wrap-events.js (hunk 18 above)
Comment (cwli24) [reply packages/browser-agent-core/common/wrap/wrap-events.js:81]: Yeah you're right; good spotting. Same as the add-event-listener.

File: packages/browser-agent-core/common/window/load.js (hunk 15 above)
Comment (metal-messiah) [reply packages/browser-agent-core/common/window/load.js:13]: To be clear, the functions you linked, `onWindowLoad()` and `onDOMContentLoaded()`, are just event listeners that early resolve if the event has already happened.  The behavior you are referring to (setting timings, now(), mark, etc), should be handled by the callers (in the page-view-event module, for instance) and not in these functions, as these are generic.  What these functions could do however, is return a flag that lets the caller know that it resolved because it had already triggered, letting the callers know not to implement timings or whatnot.

Comment (metal-messiah) [reply packages/browser-agent-core/common/window/load.js:13]: (its also used by the CDN to inject the aggregator files)

File: packages/browser-agent-core/common/window/load.js (hunk 16 above)
Comment (metal-messiah) [reply packages/browser-agent-core/common/window/load.js:19]: https://github.com/newrelic/newrelic-browser-agent/pull/255#discussion_r934991359

File: packages/browser-agent-core/common/wrap/index.js (hunk 17 above)
Comment (metal-messiah) [reply packages/browser-agent-core/common/wrap/index.js:1]: remnant of Martin's work, but serves as a single export file (imports a lot of different files, and exports them all, for ease of use)

File: packages/browser-agent-core/common/wrap/wrap-events.js (hunk 18 above)
Comment (metal-messiah) [reply packages/browser-agent-core/common/wrap/wrap-events.js:81]: only thing it does is shortcut accessing the namespaced ee without actually wrapping anything (and without needing to know which namespace belongs to which wrap module), (ie. ee.get('events'), ee.get('timings), etc).  Definitely not needed.

File: packages/browser-agent-core/common/wrap/wrap-promise.js (hunk 21 above)
Comment (metal-messiah) [reply packages/browser-agent-core/common/wrap/wrap-promise.js:66]: I believe this was added by @cwli24 for the circular interaction ticket?

File: packages/browser-agent-core/common/wrap/wrap-xhr.js (hunk 22 above)
Comment (metal-messiah) [reply packages/browser-agent-core/common/wrap/wrap-xhr.js:7]: good catch

File: packages/browser-agent-core/common/url/protocol.js (hunk 23 above)
Comment (metal-messiah) [reply packages/browser-agent-core/common/url/protocol.js:16]: What is the risk?

Comment (metal-messiah) [reply packages/browser-agent-core/common/url/protocol.js:16]: (this is implemented in the metrics feature)

File: packages/browser-agent-core/common/util/obfuscate.js (hunk 24 above)
Comment (metal-messiah) [reply packages/browser-agent-core/common/util/obfuscate.js:17]: This is implemented in the metrics feature

**Comment by aubreymasten** [comment]
Integrated into: PR https://github.com/newrelic/newrelic-browser-agent/pull/259
